browser/brave_rewards/checkout_dialog.cc

```diff
@@ -0,0 +1,343 @@
+    payments::mojom::PaymentDetailsPtr details =
+        payments::mojom::PaymentDetails::New();
+    details->id = value;
+    request->spec()->UpdateWith(std::move(details));
+    request->Pay();
+    return;
+  }
+  request->OnError(PaymentErrorReason::UNKNOWN, errors::kBatTransactionFailed);
+}
+
+void OnGetPublisherDetailsCallback(base::WeakPtr<PaymentRequest> request,
+                                   WebContents* contents,
+                                   const ledger::type::Result result,
+                                   ledger::type::PublisherInfoPtr info) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  double total = 0;
+  if (!request) {
+    return;
+  }
+
+  if (!info || info->status == ledger::type::PublisherStatus::NOT_VERIFIED) {
+    request->OnError(PaymentErrorReason::NOT_SUPPORTED,
+                     brave_rewards::errors::kInvalidPublisher);
+    return;
+  }
+
+  base::WeakPtr<payments::PaymentRequestSpec> spec = request->spec();
+  if (!spec) {
+    request->OnError(PaymentErrorReason::INVALID_DATA_FROM_RENDERER,
+                     errors::kInvalidData);
+    return;
+  }
+
+  DCHECK(base::StringToDouble(spec->details().total->amount->value, &total));
+  base::Value order_info(base::Value::Type::DICTIONARY);
+  order_info.SetDoubleKey("total", total);
+
+  base::Value params(base::Value::Type::DICTIONARY);
+  params.SetKey("orderInfo", std::move(order_info));
+
+  ShowConstrainedWebDialogWithAutoResize(
+      contents->GetBrowserContext(),
+      std::make_unique<CheckoutDialogDelegate>(std::move(params), request),
+      contents, gfx::Size(kDialogWidth, kDialogMinHeight),
+      gfx::Size(kDialogWidth, kDialogMaxHeight));
+}
+
+}  // namespace
+
+enum DialogCloseReason {
+  Complete,
+  InsufficientBalance,
+  UnverifiedWallet,
+  UserCancelled
+};
+
+CheckoutDialogDelegate::CheckoutDialogDelegate(
+    base::Value params,
+    base::WeakPtr<PaymentRequest> request)
+    : params_(std::move(params)), request_(request) {}
+
+CheckoutDialogDelegate::~CheckoutDialogDelegate() = default;
+
+ui::ModalType CheckoutDialogDelegate::GetDialogModalType() const {
+  return ui::MODAL_TYPE_WINDOW;
+}
+
+base::string16 CheckoutDialogDelegate::GetDialogTitle() const {
+  return base::string16();
+}
+
+GURL CheckoutDialogDelegate::GetDialogContentURL() const {
+  return GURL(kBraveUICheckoutURL);
+}
+
+void CheckoutDialogDelegate::GetWebUIMessageHandlers(
+    std::vector<WebUIMessageHandler*>* handlers) const {
+  handlers->push_back(new CheckoutDialogHandler(request_));
+}
+
+void CheckoutDialogDelegate::GetDialogSize(gfx::Size* size) const {}
+
+std::string CheckoutDialogDelegate::GetDialogArgs() const {
+  std::string json;
+  base::JSONWriter::Write(params_, &json);
+  return json;
+}
+
+void CheckoutDialogDelegate::OnDialogClosed(const std::string& result) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  int reason = 0;
+  DCHECK(base::StringToInt(result, &reason));
```

Comment:
got a crash here: closed the dialog (showing "Sorry... only verified wallets etc.") with Esc
```
[1643397:1643397:0416/182421.529374:FATAL:checkout_dialog.cc(145)] Check failed: base::StringToInt(result, &reason). 
#0 0x7f57096fcb19 base::debug::CollectStackTrace()
#1 0x7f57095fcb33 base::debug::StackTrace::StackTrace()
#2 0x7f570961c523 logging::LogMessage::~LogMessage()
#3 0x7f570961cece logging::LogMessage::~LogMessage()
#4 0x556d39124f71 brave_rewards::CheckoutDialogDelegate::OnDialogClosed()
#5 0x556d3a926c61 (anonymous namespace)::ConstrainedDialogWebView::WindowClosing()
#6 0x7f5701b47e0d views::Widget::OnNativeWidgetDestroying()
#7 0x7f5701b6f27b views::NativeWidgetAura::OnWindowDestroying()
#8 0x7f5702a73b19 aura::Window::~Window()
#9 0x7f5702a746ae aura::Window::~Window()
#10 0x7f5701a377d4 base::internal::Invoker<>::RunOnce()
```

Reply:
closing the tab also causes a DCHECK failure